### PR TITLE
RDRP-650: Auto outlier unit tests

### PR DESCRIPTION
**Pull request by JenCheshire**
Unit tests for following `auto_outliers.py` functions:

**`validate_config()`**
- Test for warning where both clips are zero
- Test for error where a clip is a minus

**`filter_valid()`**
- Test to ensure filtering is correct

**`flag_outliers()`**
- Test to check flagging correct
- Five tests to check rounding is behaving as expected: clip the same, different size df's
- Test that nothing is flagged when clips = 0

**`decide_outliers()`**
- Test to ensure decide outliers working correctly

**`log_outlier_info`**
- Test to ensure logging information correct

**`run_auto_flagging()`**
- Test running whole auto_outlier process and checking that if there is one outlier value, the whole record is flagged as an outlier

**`apply_short_form_filters()`**
- Test to ensure short form filtering is correct

### CODE REVIEW

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 1]
```diff
@@ -1,11 +1,102 @@
...
+class TestValidateConfigZero:
+    """Unit tests for validate_config function."""
+
+    def test_validate_config(self, caplog):
+        """Test for validate_config function."""
+
+        with caplog.at_level(logging.WARNING):
+            auto.validate_config(0.0, 0.0, ["701", "702", "703"])
+
+        assert "upper_clip and lower_clip both zero:" in caplog.text
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py:26]: Great but I think there should be 2 more tests. 1 where a non-List is pass and another where the upper and/or lower clip is =>1

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 2]
```diff
@@ -1,11 +1,102 @@
...
+
+    def test_filter_valid(self):
+        """Test for filter_valid function."""
+        input_df = self.create_input_df()
+        expected_df = self.create_expected_df()
+
+        value_col = "701"
+        result_df = auto.filter_valid(input_df, value_col)
+
+        assert_frame_equal(result_df.reset_index(drop=True), expected_df.reset_index(drop=True))
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py:126]: I think this is great and covers all of the conditions. 

Should there be another test to make sure the logger and ValueError are raised when there's an empty dataframe?

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 3]
```diff
@@ -100,13 +191,549 @@ def test_flag_outliers(self):
...
+
+    def test_flag_outliers(self):
+        """Test for flag_outliers function."""
+        input_df = self.create_input_df()
+        expected_df = self.create_expected_df()
+
+        upper_clip = 0.05
+        lower_clip = 0
+        value_col = "701"
+        result_df = auto.flag_outliers(input_df, upper_clip, lower_clip, value_col)
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py:465]: This looks good to me. I have also checked it mathematically in Excel

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 4]
```diff
@@ -100,13 +191,549 @@ def test_flag_outliers(self):
...
+
+        assert_frame_equal(result_df, expected_df)
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py]: This has been checked in Excel.

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 5]
```diff
@@ -100,13 +191,549 @@ def test_flag_outliers(self):
...
+
+    def test_flag_outliers(self):
+        """Test for flag_outliers function."""
+        input_df = self.create_input_df()
+        expected_df = self.create_expected_df()
+
+        upper_clip = 0.05
+        lower_clip = 0
+        value_col = "701"
+        result_df = auto.flag_outliers(input_df, upper_clip, lower_clip, value_col)
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py:723]: Tested in Excel and it's fine.

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 6]
```diff
@@ -100,13 +191,549 @@ def test_flag_outliers(self):
...
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py:867]: Again, testing in Excel reveals it's fine.

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 7]
```diff
@@ -155,6 +782,251 @@ def test_decide_outliers(self):
...
+
+    def test_log_outlier_info(self, caplog):
+        """Test for log_outlier_info function."""
+
+        input_df = self.create_input_df()
+
+        with caplog.at_level(logging.INFO):
+            auto.log_outlier_info(input_df, "value_col")
+
+        assert "1 outliers were detected out of a total of 2 valid entries in column value_col" in caplog.text
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py:970]: Checked this thoroughly and it's fine

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 8]
```diff
@@ -155,6 +782,251 @@ def test_decide_outliers(self):
...
+        """Test for flag_outliers function."""
+        input_df = self.create_input_df()
+        expected_df = self.create_expected_df()
+
+        form_type_no: str = "0006"
+        sel_type: str = "P"
+
+        result_df = auto.apply_short_form_filters(input_df, form_type_no, sel_type)
+
+        assert_frame_equal(result_df.reset_index(drop=True), expected_df.reset_index(drop=True))
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py]: Looks good.

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 9]
```diff
@@ -155,6 +782,251 @@ def test_decide_outliers(self):
...
+        expected_df = self.create_expected_df()
+
+
+        upper_clip = 0.05
+        lower_clip = 0.05
+        flag_cols = ["701", "702", "703", "704"]
+
+        result_df = auto.run_auto_flagging(input_df, upper_clip, lower_clip, flag_cols)
+
+        assert_frame_equal(result_df, expected_df)
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py:1064]: The code looks good and I have tested it in Excel. It all checks out.

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 10]
```diff
@@ -155,6 +782,251 @@ def test_decide_outliers(self):
...
+
+    def test_flag_outliers(self):
+        """Test for flag_outliers function."""
+        input_df = self.create_input_df()
+        expected_df = self.create_expected_df()
+
+        upper_clip = 0
+        lower_clip = 0
+        value_col = "701"
+        result_df = auto.flag_outliers(input_df, upper_clip, lower_clip, value_col)
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py]: Checked this by eye (not in Excel). It looks good.

File: tests/test_outlier_detection/test_auto_outliers.py [hunk 11]
```diff
@@ -155,6 +782,251 @@ def test_decide_outliers(self):
         expected_df = self.create_expected_df()
```
Comment (jwestw) [inline tests/test_outlier_detection/test_auto_outliers.py:923]: Checked the `TestDecideOutliers` too even though it's not part of the PR. Looks good.

**Review by jwestw — APPROVED** [review]
I think we need a unit test to show `_normal_round` is working as expected. We can work out how to do this. 

Otherwise this all looks like stellar code!

**Comment by jwestw** [comment]
The Excel sheet I made is available to team members internally [on Sharepoint](https://officenationalstatistics-my.sharepoint.com/:x:/g/personal/james_westwood_ons_gov_uk/EZ_oRmoKjPBEhks7cemymqgBRG1_EduMbFvJznxzUbNkxQ?e=941Kff)

File: tests/test_outlier_detection/test_auto_outliers.py (hunk 1 above)
Comment (JenCheshire) [reply tests/test_outlier_detection/test_auto_outliers.py:26]: Tests added

File: tests/test_outlier_detection/test_auto_outliers.py (hunk 2 above)
Comment (JenCheshire) [reply tests/test_outlier_detection/test_auto_outliers.py:126]: Test added

**Comment by JenCheshire** [comment]
Hi @jwestw - I have added those tests, and taken `normal_rounding()` out of the `flag_outliers()` function, and created a standalone test for it. Please could you have a quick look at those changes.  Thanks

File: src/outlier_detection/auto_outliers.py [hunk 12]
```diff
@@ -99,22 +99,6 @@ def flag_outliers(
     # Filter for valid sampled data and positive values in the value column
     filtered_df = filter_valid(df, value_col)
 
-    def _normal_round(x: float) -> int:
```
Comment (zorge69) [inline src/outlier_detection/auto_outliers.py:102]: I called a single leading underscore to signify that it's an internal use function. Yet, if you are going to unit test it, you will need to import it, so it is correct to remove the underscore.

**Review by zorge69 — APPROVED** [review]

**Review by jwestw — APPROVED** [review]
The changes look good. All tests pass.
